pretraining/utils/dataset_utils.py

```diff
@@ -820,6 +820,13 @@ def load_state_dict(self, state_dicts, sharded_input=False):
         ), "Streaming_Doc_Dataset does not support rescaling"
         return super().load_state_dict(state_dicts, sharded_input)
 
+@runtime_checkable
+class _With_Datasets(Protocol):
+    datasets: List[Any]
+
+@runtime_checkable
+class _With_Docset(Protocol):
+    docset: List[Any]
```

Comment:
Alternatives could be to add `datasets` and `docset` to `_Stateful_Dataset` or to create an intermediate base class with those two attributes. This looked like the least intrusive option for now, but let me know if I should change it.